### PR TITLE
Fix #583: NSUserdefaults synchronize

### LINKED ISSUE #583: NSUserDefaults are not being saved across launches
User defaults settings are not being restored when re-launching the app.

Running the following code should output `Test value: 0` the first time, but `Test value: 1` for every subsequent run of the app. However, it always outputs `Test value: 0`.

```objc
NSLog(@"Test value: %d", [[NSUserDefaults standardUserDefaults] boolForKey:@"Test"]);
[[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"Test"];
```

### PR DESCRIPTION
**Pull request by jofre-ms**
* NSUserdefaults synchronize must be called after app settings have changed.
* Added boolean to NSUSerDefaults WOCCatalog page

### CODE REVIEW

File: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m [hunk 1]
```diff
@@ -51,7 +52,19 @@ - (void)viewDidLoad {
...
+    // Slider for retrieving favorite bool
+    UISwitch* switchFavoriteBool = [[UISwitch alloc] initWithFrame:frame];
+    [switchFavoriteBool setOn:[[standardDefaults objectForKey:@"FavoriteBool"] boolValue] animated:NO];
+    [_textFields addObject:switchFavoriteBool];
+    [switchFavoriteBool addTarget:self action:@selector(changeFavoriteBool) forControlEvents:UIControlEventValueChanged];
+
+    [self tableView].allowsSelection = NO;    
+}
+
+- (void) changeFavoriteBool {
```
Comment (DHowett-MSFT) [inline samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m]: nit: formatting

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 2]
```diff
@@ -58,6 +59,9 @@ - (instancetype)init {
 
     _registrationDict = [NSMutableDictionary dictionary];
 
+    _synchronizeQueue = [NSOperationQueue new];
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: This will leak, as `+new` returns a **+1**.

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 3]
```diff
@@ -48,6 +48,7 @@
 
 @implementation NSUserDefaults {
     StrongId<NSMutableDictionary> _registrationDict;
+    StrongId<NSOperationQueue> _synchronizeQueue;
 }
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: I'm somewhat wary of introducing a dependency on NSOperationQueue here; there might be some way to do this with dispatch groups or another primitive.
NSOperationQueue could spawn its own thread pool to service requests like this, but we don't need an entire thread sitting around dedicated to saving user preferences.

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 4]
```diff
@@ -245,6 +249,12 @@ - (void)setObject:(id)value forKey:(NSString*)key {
     CFPreferencesSetAppValue(static_cast<CFStringRef>(key), valueCopy, kCFPreferencesCurrentApplication);
 
     [[NSNotificationCenter defaultCenter] postNotificationName:NSUserDefaultsDidChangeNotification object:self];
+
+    if ([_synchronizeQueue operationCount] < 2) {
+        [_synchronizeQueue addOperationWithBlock:^void(void) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: It might be worth calling out why we're using 2 here.

**Comment by DHowett-MSFT** [comment]
:clock1:

**Comment by jofre-ms** [comment]
@bbowman is added to the review. #Closed

File: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m [hunk 5]
```diff
@@ -30,12 +30,13 @@ - (void)viewDidLoad {
     [super viewDidLoad];
 
     NSUserDefaults* standardDefaults = [NSUserDefaults standardUserDefaults];
-    NSDictionary* dict = @{@"FavoriteWord" : @"Blue"};
+    NSDictionary* dict = @{@"FavoriteWord" : @"Blue", @"FavoriteBool" : [NSNumber numberWithBool:YES]};
```
Comment (rajsesh-msft) [inline samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m]: >FavoriteBool [](start = 56, length = 12)

why this change?

File: Frameworks/Foundation/NSUserDefaults.mm (hunk 3 above)
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSUserDefaults.mm]: the proper implementation of NSOperationQueue would use thread pools, but agree that our current implementation is not efficient.  

---
In reply to: [69786383](https://github.com/Microsoft/WinObjC/pull/612#discussion_r69786383) [](ancestors = 69786383)

**Comment by rajsesh-msft** [comment]
:clock1:

File: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m [hunk 6]
```diff
@@ -51,7 +52,19 @@ - (void)viewDidLoad {
 
     [_textFields addObject:textField];
 
-    [self tableView].allowsSelection = NO;
+    // Slider for retrieving favorite bool
+    UISwitch* switchFavoriteBool = [[UISwitch alloc] initWithFrame:frame];
+    [switchFavoriteBool setOn:[[standardDefaults objectForKey:@"FavoriteBool"] boolValue] animated:NO];
```
Comment (jofre-ms) [inline samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m]: boolForKey

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 7]
```diff
@@ -245,6 +249,12 @@ - (void)setObject:(id)value forKey:(NSString*)key {
     CFPreferencesSetAppValue(static_cast<CFStringRef>(key), valueCopy, kCFPreferencesCurrentApplication);
```
Comment (rajsesh-msft) [inline Frameworks/Foundation/NSUserDefaults.mm]: >CFPreferencesSetAppValue [](start = 4, length = 24)

we need to synchronize of the apps use CFPreferences* apis to set values.

Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: On the contrary, the documentation for  CFPreferencesSetAppValue states "You must call the CFPreferencesAppSynchronize function in order for your changes to be saved to permanent storage." - Therefore this automatic synchronize logic only needs to exist in NSUserDefaults.

---
In reply to: [69979853](https://github.com/Microsoft/WinObjC/pull/612#discussion_r69979853) [](ancestors = 69979853)

File: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m (hunk 5 above)
Comment (jofre-ms) [inline samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m]: Testing a bool strictly increases our test coverage. I wrote these changes because I couldn't answer the question of whether this scenario had regressed.

---
In reply to: [69818261](https://github.com/Microsoft/WinObjC/pull/612#discussion_r69818261) [](ancestors = 69818261)

File: Frameworks/Foundation/NSUserDefaults.mm (hunk 2 above)
Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: _synchronizeQueue is a StrongId, which will release the object at the appropriate time.

---
In reply to: [69786087](https://github.com/Microsoft/WinObjC/pull/612#discussion_r69786087) [](ancestors = 69786087)

File: Frameworks/Foundation/NSUserDefaults.mm (hunk 3 above)
Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: A quick examination of our NSOperationQueue implementation shows that we destroy the worker thread whenever the queue is empty and create a new thread when an operation is added. That's inefficient compared to using threadpools, but we should make the fix in NSOperationQueue rather than switching to a threadpool in NSUserDefaults because:

1) The performance gains in NSUserDefaults would be nominal compared to the already huge cost of writing to file
2) We would now need to write two fixes instead of one and maintain two threadpool/queue implementations instead of one.

---
In reply to: [69818599](https://github.com/Microsoft/WinObjC/pull/612#discussion_r69818599) [](ancestors = 69818599,69786383)

File: Frameworks/Foundation/NSUserDefaults.mm (hunk 2 above)
Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: Looks like I need .attach

---
In reply to: [69994676](https://github.com/Microsoft/WinObjC/pull/612#discussion_r69994676) [](ancestors = 69994676,69786087)

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 8]
```diff
@@ -245,6 +249,12 @@ - (void)setObject:(id)value forKey:(NSString*)key {
...
 
     [[NSNotificationCenter defaultCenter] postNotificationName:NSUserDefaultsDidChangeNotification object:self];
+
+    if ([_synchronizeQueue operationCount] < 2) {
```
Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: Maybe this isn't the best way to determine if an operation should be added to the queue. I'm going to experiment with alternate designs and see if it makes my perf test any faster.

**Comment by rajsesh-msft** [comment]
>     textField.textAlignment = UITextAlignmentRight;

Unrelated to this change, but this layout causes this page to be unusable on phone.  Can you make the text field left aligned? 

---
Refers to: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m:49 in a37872a. [](commit_id = a37872a2d97035c48efb8527320124d8da4f5164, deletion_comment = False)

**Comment by rajsesh-msft** [comment]
>     textField.backgroundColor = [UIColor blueColor];

can we just use the defaults instead?

---
Refers to: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m:43 in a37872a. [](commit_id = a37872a2d97035c48efb8527320124d8da4f5164, deletion_comment = False)

File: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m [hunk 9]
```diff
@@ -30,12 +30,13 @@ - (void)viewDidLoad {
     [super viewDidLoad];
 
     NSUserDefaults* standardDefaults = [NSUserDefaults standardUserDefaults];
-    NSDictionary* dict = @{@"FavoriteWord" : @"Blue"};
+    NSDictionary* dict = @{@"FavoriteWord" : @"Blue", @"FavoriteBool" : [NSNumber numberWithBool:YES]};
     [standardDefaults registerDefaults:dict];
     NSString* placeHolder = [NSString stringWithFormat:@"Favorite Word: %@ (Type a new value here and relaunch!)", [standardDefaults stringForKey:@"FavoriteWord"]];
```
Comment (rajsesh-msft) [inline samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m]: >Favorite Word: %@ (Type a new value here and relaunch!) [](start = 57, length = 55)

This string will not fit on phone, please use the current value as the placeholder.

File: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m [hunk 10]
```diff
@@ -30,28 +30,41 @@ - (void)viewDidLoad {
...
     _textFields = [NSMutableArray new];
 
+    // Text box for retrieving favorite string
     CGRect frame = CGRectMake(c_originX, c_originY, c_width, c_height);
     UITextField* textField = [[UITextField alloc] initWithFrame:frame];
-    textField.textColor = [UIColor blackColor];
-    textField.backgroundColor = [UIColor blueColor];
-    textField.secureTextEntry = NO;
-    textField.font = [UIFont systemFontOfSize:17.0];
+//    textField.textColor = [UIColor blackColor];
```
Comment (jofre-ms) [inline samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m]: whoops, need to remove these

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 11]
```diff
@@ -56,14 +59,23 @@ @implementation NSUserDefaults {
 - (instancetype)init {
     [super init];
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSUserDefaults.mm]: > [super init]; [](start = 3, length = 14)

if(self = [super init])

Comment (msft-Jeyaram) [inline Frameworks/Foundation/NSUserDefaults.mm]: the call to super should be through the designated initializer.

File: samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m [hunk 12]
```diff
@@ -30,28 +30,35 @@ - (void)viewDidLoad {
...
-    textField.textAlignment = UITextAlignmentRight;
-    textField.spellCheckingType = UITextSpellCheckingTypeNo;
+    textField.textAlignment = UITextAlignmentLeft;
     textField.delegate = self;
 
     [_textFields addObject:textField];
 
-    [self tableView].allowsSelection = NO;
+    // Slider for retrieving favorite bool
+    UISwitch* switchFavoriteBool = [[UISwitch alloc] initWithFrame:frame];
```
Comment (msft-Jeyaram) [inline samples/WOCCatalog/WOCCatalog/UserDefaultsViewController.m]: >switchFavoriteBool [](start = 14, length = 18)

needs to be released

**Comment by msft-Jeyaram** [comment]
:clock1:

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 13]
```diff
@@ -119,16 +132,44 @@ - (void)setPersistentDomain:(id)domain forName:(NSString*)name {
  @Status Interoperable
 */
 - (BOOL)synchronize {
-    return CFPreferencesAppSynchronize(kCFPreferencesCurrentApplication);
+    _cacheLock.lock();
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: prefer `std::lock_guard` for scoped locking around mutices

Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: We need fine control over the lock here, as CFPreferencesAppSynchronize should be called outside of _cacheLock.

---
In reply to: [70540876](https://github.com/Microsoft/WinObjC/pull/612#discussion_r70540876) [](ancestors = 70540876)

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 14]
```diff
@@ -119,16 +132,44 @@ - (void)setPersistentDomain:(id)domain forName:(NSString*)name {
  @Status Interoperable
```
Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: >Interoperable [](start = 9, length = 13)

this should really be caveat - with a note that synchronize does not read from file

File: Frameworks/Foundation/NSUserDefaults.mm (hunk 13 above)
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: Fortunately, `std::lock_guard` locks a scope, and you could introduce a scope for this.

---
In reply to: [70674087](https://github.com/Microsoft/WinObjC/pull/612#discussion_r70674087) [](ancestors = 70674087,70540876)

Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: ok then let's do it everywhere.

---
In reply to: [70694534](https://github.com/Microsoft/WinObjC/pull/612#discussion_r70694534) [](ancestors = 70694534,70674087,70540876)

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 15]
```diff
@@ -47,40 +47,68 @@
...
 */
 + (NSUserDefaults*)standardUserDefaults {
-    static NSUserDefaults* standard;
-
-    if (standard == nil) {
-        standard = [self new];
+    if (_standard == nil) {
+        _standard = [self new];
     }
```
Comment (mnithish) [inline Frameworks/Foundation/NSUserDefaults.mm]: not thread safe

File: Frameworks/UIKit/UIApplicationMain.mm [hunk 16]
```diff
@@ -320,6 +320,14 @@ void _UIApplicationShutdown() {
     [[UIApplication sharedApplication] _sendHighMemoryWarning];
 }
 
+extern "C" void UIApplicationMainHandleSuspendEvent() {
+    [[UIApplication sharedApplication] _sendSuspendEvent];
```
Comment (mnithish) [inline Frameworks/UIKit/UIApplicationMain.mm]: >_sendSuspendEvent [](start = 39, length = 17)

nit: we are not sending any suspend event (unlike other methods) we are handling it.

Comment (mnithish) [inline Frameworks/UIKit/UIApplicationMain.mm]: Same for resume below

---
In reply to: [70830139](https://github.com/Microsoft/WinObjC/pull/612#discussion_r70830139) [](ancestors = 70830139)

File: docs/Foundation/NSUserDefaults.md [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+
+## **DEV DESIGN SPECIFICATION: NSUserDefaults
+
+### Feature Summary
+
+NSUserDefaults is a Foundation interface to CFPreferences that stores a preference plist in the package appdata directory.
+
+### Technology/Architecture Decisions and Caveats
+
+- Changes to standardDefaults are written to cache and synchronized to disk as soon as possible on an asyncrhonous queue.
```
Comment (mnithish) [inline docs/Foundation/NSUserDefaults.md]: >asyncrhonous [](start = 102, length = 12)

nit

**Comment by mnithish** [comment]
:clock1:

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 18]
```diff
@@ -47,40 +47,68 @@
...
+
 /**
  @Status Interoperable
 */
 + (NSUserDefaults*)standardUserDefaults {
-    static NSUserDefaults* standard;
-
-    if (standard == nil) {
-        standard = [self new];
+    if (_standard == nil) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: I think you should keep this as-is; use a magic static to force threadsafe init.
```
static NSUserDefaults* standard = [self new];
return standard;
```

C++11 ensures it only happens once.

Ideally, the suspend/resume methods would be per-instance; UIApplication would then call `[[NSUserDefaults standardUserDefaults] _zhuLiDoTheThing]`.

Comment (jofre-ms) [inline Frameworks/Foundation/NSUserDefaults.mm]: The reason I avoided the [NSUserDefaults standardUserDefaults] call is to prevent initialization of standardUserDefaults when we suspend, if it hasn't already been initialized. I'll try a different workaround.

---
In reply to: [70878479](https://github.com/Microsoft/WinObjC/pull/612#discussion_r70878479) [](ancestors = 70878479,70878446)

File: Frameworks/Foundation/NSUserDefaults.mm [hunk 19]
```diff
@@ -232,8 +302,32 @@ - (double)doubleForKey:(NSString*)defaultName {
...
+}
+
+- (NSOperationQueue*) _synchronizeQueue {
+    return _synchronizeQueue;
+}
+
+- (void) _suspendSynchronize {
+    [_synchronizeQueue setSuspended:YES];
+    [_synchronizeQueue waitUntilAllOperationsAreFinished];
+}
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: What happens if we suspend it, then another thread injects a synchronize, and then we try to wait on an operation that'll never complete (since it was suspended)?

Comment (mnithish) [inline Frameworks/Foundation/NSUserDefaults.mm]: That should not a problem because, _suspendSynchronize is called just before the application is being suspended, so the other thread will also get suspended and when the application resumes _resumeSynchronize will get called on the UI thread and that will resume the other thread from the wait.

---
In reply to: [70900447](https://github.com/Microsoft/WinObjC/pull/612#discussion_r70900447) [](ancestors = 70900447)

Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSUserDefaults.mm]: Discussed offline; Doing this should suffice:
```
[_synchronizeQueue addOperations:@[[NSBlockOperation blockOperationWithBlock:^{ [_synchronizeQueue suspend]; }]] waitUntilFinished:YES];
```

---
In reply to: [70908191](https://github.com/Microsoft/WinObjC/pull/612#discussion_r70908191) [](ancestors = 70908191,70900447)

**Comment by DHowett-MSFT** [comment]
In CI build.
